radixdlt/src/main/java/com/radixdlt/consensus/PendingVotes.java

```diff
@@ -41,26 +81,74 @@ public PendingVotes(Hasher hasher) {
 
 	/**
 	 * Inserts a vote for a given vertex, attempting to form a quorum certificate for that vertex.
-	 *
+	 * <p>
 	 * A QC will only be formed if permitted by the {@link ValidatorSet}.
+	 *
 	 * @param vote The vote to be inserted
 	 * @return The generated QC, if any
 	 */
 	public Optional<QuorumCertificate> insertVote(Vote vote, ValidatorSet validatorSet) {
-		final Hash voteHash = hasher.hash(vote.getVoteData());
-
-		ECDSASignature signature = vote.getSignature().orElseThrow(() -> new IllegalArgumentException("vote is missing signature"));
-		ValidationState validationState = votes.computeIfAbsent(voteHash, validatorSet::newValidationState);
-
-		// try to form a QC with the added signature according to the requirements
-		if (!validationState.addSignature(vote.getAuthor(), signature)) {
-			// if no QC could be formed, update pending and return nothing
-			return Optional.empty();
-		} else {
-			// if QC could be formed, remove pending and return formed QC
-			votes.remove(voteHash);
-			QuorumCertificate qc = new QuorumCertificate(vote.getVoteData(), validationState.signatures());
-			return Optional.of(qc);
+		final ECPublicKey voteAuthor = vote.getAuthor();
+		final VoteData voteData = vote.getVoteData();
+		final Hash voteHash = this.hasher.hash(voteData);
+		final ECDSASignature signature = vote.getSignature().orElseThrow(() -> new IllegalArgumentException("vote is missing signature"));
+		// Only process for valid validators and signatures
+		if (validatorSet.containsKey(voteAuthor) && voteAuthor.verify(voteHash, signature)) {
+			final View voteView = voteData.getProposed().getView();
+			if (replacePreviousVote(voteAuthor, voteView, voteHash)) {
+				// If there is no equivocation or duplication, we process the vote.
+				ValidationState validationState = this.voteState.computeIfAbsent(voteHash, k -> validatorSet.newValidationState());
+
+				// try to form a QC with the added signature according to the requirements
+				if (validationState.addSignature(voteAuthor, signature) && validationState.complete()) {
+					// QC can be formed, so return it
+					QuorumCertificate qc = new QuorumCertificate(vote.getVoteData(), validationState.signatures());
+					return Optional.of(qc);
+				}
+			}
+		}
+		// No QC could be formed, so return nothing
+		return Optional.empty();
+	}
+
+	private boolean replacePreviousVote(ECPublicKey author, View voteView, Hash voteHash) {
+		PreviousVote thisVote = new PreviousVote(voteView, voteHash);
+		PreviousVote previousVote = this.previousVotes.put(author, thisVote);
+		if (previousVote == null) {
+			// No previous vote for this author, all good here
+			return true;
+		}
+
+		if (thisVote.equals(previousVote)) {
+			// Just going to ignore this duplicate vote for now.
+			// However, we can't count duplicate votes multiple times.
+			return false;
 		}
+
+		// Prune last pending vote from the pending votes.

```

Comment:
Can we add a comment here in that this inherently limits the number of pending vertices which are in the pipeline to create a QC (aka mechanism in Event-Driven Hotstuff)